examples/neutra.py

```diff
@@ -0,0 +1,210 @@
+"""
+This example illustrates the use of `NeuTraReparam` to run neural transport HMC [1]
+on a toy model that draws from a banana-shaped bivariate distribution [2]. We first
+train an autoguide by using `AutoNormalizingFlow` that learns a transformation from
+a simple latent space (isotropic gaussian) to the more complex geometry of the
+posterior. Subsequently, we use `NeuTraReparam` to run HMC and draw samples from this
+simplified "warped" posterior. Finally, we use our learnt transformation to transform
+these samples back to the original space. For comparison, we also draw samples from
+a NeuTra-reparametrized model that uses a much simpler `AutoDiagonalNormal` guide.
+
+References:
+----------
+[1] Hoffman, M., Sountsov, P., Dillon, J. V., Langmore, I., Tran, D., and Vasudevan,
+ S. Neutra-lizing bad geometry in hamiltonian monte carlo using neural transport.
+ arXiv preprint arXiv:1903.03704, 2019.
+[2] Wang Z., Broccardo M., and Song J. Hamiltonian Monte Carlo Methods for Subset
+ Simulation in Reliability Analysis. arXiv preprint arXiv:1706.01435, 2018.
+"""
+
+import argparse
+import logging
+from functools import partial
+
+import torch
+import matplotlib.pyplot as plt
+import seaborn as sns
+from matplotlib.gridspec import GridSpec
+
+import pyro
+from pyro import optim, poutine
+import pyro.distributions as dist
+from pyro.distributions import constraints
+from pyro.distributions.transforms import iterated, block_autoregressive
+from pyro.infer import MCMC, NUTS, SVI, Trace_ELBO
+from pyro.infer.autoguide import AutoDiagonalNormal, AutoNormalizingFlow
+from pyro.infer.reparam import NeuTraReparam
+
+
+logging.basicConfig(format='%(message)s', level=logging.INFO)
+
+
+class BananaShaped(dist.TorchDistribution):
+    support = constraints.real_vector
+
+    def __init__(self, a, b, rho=0.9):
+        self.a = a
+        self.b = b
+        self.rho = rho
+        self.mvn = dist.MultivariateNormal(torch.tensor([0., 0.]),
+                                           covariance_matrix=torch.tensor([[1., self.rho], [self.rho, 1.]]))
+        super().__init__(event_shape=(2,))
+
+    def sample(self, sample_shape=()):
+        u = self.mvn.sample(sample_shape)
+        u0, u1 = u[..., 0], u[..., 1]
+        a, b = self.a, self.b
+        x = a * u0
+        y = (u1 / a) + b * (u0 ** 2 + a ** 2)
+        return torch.stack([x, y], -1)
+
+    def log_prob(self, x):
```

Comment:
this is a properly normalized probability?

Reply:
It should be since the determinant of the jacobian is 1.